File: spendingtapper/app/src/main/java/dev/xsk1d/spendingtapper/io/Csv.kt
package dev.xsk1d.spendingtapper.io

import dev.xsk1d.spendingtapper.data.Expense
import dev.xsk1d.spendingtapper.data.Kind
import dev.xsk1d.spendingtapper.domain.Money
import java.io.BufferedReader
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Export and import in plain RFC 4180 CSV, so the file opens in any spreadsheet
 * and a description containing a comma, a quote or a newline still survives the
 * round trip.
 */
object Csv {

    val HEADER = listOf("id", "occurred_at", "amount", "kind", "description", "with_who")

    private val TIMESTAMP: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")

    data class ImportResult(
        val expenses: List<Expense>,
        val skipped: Int,
    )

    fun export(expenses: List<Expense>, zone: ZoneId): String = buildString {
        appendLine(HEADER.joinToString(",") { escape(it) })
        for (e in expenses) {
            val local = LocalDateTime.ofInstant(Instant.ofEpochMilli(e.occurredAt), zone)
            val row = listOf(
                e.id.toString(),
                local.format(TIMESTAMP),
                Money.formatCents(e.amountCents),
                e.kind.name,
                e.description,
                e.withWho,
            )
            appendLine(row.joinToString(",") { escape(it) })
        }
    }

    fun import(reader: BufferedReader, zone: ZoneId, now: Long): ImportResult {
        val rows = parse(reader.readText())
        if (rows.isEmpty()) return ImportResult(emptyList(), 0)

        // Tolerate a file with or without a header line.
        val first = rows.first().map { it.trim().lowercase() }
        val body = if (first.getOrNull(0) == "id" || first.getOrNull(2) == "amount") rows.drop(1) else rows

        val parsed = mutableListOf<Expense>()
        var skipped = 0
        for (row in body) {
            val expense = parseRow(row, zone, now)
            if (expense == null) skipped++ else parsed += expense
        }
        return ImportResult(parsed, skipped)
    }

    private fun parseRow(row: List<String>, zone: ZoneId, now: Long): Expense? {
        if (row.size < 4) return null
        if (row.all { it.isBlank() }) return null

        val amountCents = Money.parseAmount(row[2]) ?: return null
        val occurredAt = parseTimestamp(row[1], zone) ?: return null
        val kind = runCatching { Kind.valueOf(row[3].trim().uppercase()) }.getOrNull() ?: return null

        return Expense(
            // id 0 lets Room assign a fresh one, which is what we want on import:
            // merging a backup into a non-empty database must not overwrite live rows.
            id = 0,
            amountCents = amountCents,
            kind = kind,
            description = row.getOrNull(4).orEmpty().trim(),
            occurredAt = occurredAt,
            withWho = Expense.joinPeople(Expense.splitPeople(row.getOrNull(5).orEmpty())),
            createdAt = now,
        )
    }

    private fun parseTimestamp(raw: String, zone: ZoneId): Long? {
        val text = raw.trim().replace('T', ' ').removeSuffix("Z")
        if (text.isEmpty()) return null
        val candidates = listOf(text, "$text:00", "$text 00:00:00")
        for (candidate in candidates) {
            val parsed = runCatching { LocalDateTime.parse(candidate, TIMESTAMP) }.getOrNull()
            if (parsed != null) return parsed.atZone(zone).toInstant().toEpochMilli()
        }
        // A bare epoch-millis column is also accepted.
        return text.toLongOrNull()
    }

    private fun escape(value: String): String =
        if (value.any { it == ',' || it == '"' || it == '\n' || it == '\r' }) {
            "\"" + value.replace("\"", "\"\"") + "\""
        } else {
            value
        }

    /** A small RFC 4180 reader: handles quoted fields, doubled quotes and embedded newlines. */
    fun parse(text: String): List<List<String>> {
        val rows = mutableListOf<List<String>>()
        var row = mutableListOf<String>()
        val field = StringBuilder()
        var inQuotes = false
        var i = 0

        fun endField() {
            row.add(field.toString())
            field.setLength(0)
        }

        fun endRow() {
            endField()
            if (row.size > 1 || row.first().isNotBlank()) rows.add(row)
            row = mutableListOf()
        }

        while (i < text.length) {
            val c = text[i]
            when {
                inQuotes && c == '"' && i + 1 < text.length && text[i + 1] == '"' -> {
                    field.append('"'); i++
                }
                c == '"' -> inQuotes = !inQuotes
                !inQuotes && c == ',' -> endField()
                !inQuotes && (c == '\n' || c == '\r') -> {
                    if (c == '\r' && i + 1 < text.length && text[i + 1] == '\n') i++
                    endRow()
                }
                else -> field.append(c)
            }
            i++
        }
        if (field.isNotEmpty() || row.isNotEmpty()) endRow()
        return rows
    }
}

File: src/lib/csv.ts
import { format as formatDate, isValid, parse as parseDate } from 'date-fns'
import { formatCents, parseAmount } from './money'
import { dedupePeople, joinPeople, splitPeople } from './people'
import type { Expense, Kind } from './types'

/**
 * Export and import in plain RFC 4180 CSV, so the file opens in any spreadsheet
 * and a description containing a comma, a quote or a newline still survives the
 * round trip.
 */

export const HEADER = ['id', 'occurred_at', 'amount', 'kind', 'description', 'with_who'] as const

const TIMESTAMP = 'yyyy-MM-dd HH:mm:ss'

export type ImportResult = {
  /** Parsed rows, each already given a fresh id. */
  expenses: Expense[]
  /** Rows that could not be read. Counted rather than thrown, so one bad line
   *  does not cost you the whole backup. */
  skipped: number
}

function escape(value: string): string {
  return /[",\n\r]/.test(value) ? `"${value.replace(/"/g, '""')}"` : value
}

export function exportCsv(expenses: Expense[]): string {
  const lines = [HEADER.map(escape).join(',')]
  for (const expense of expenses) {
    lines.push(
      [
        expense.id,
        formatDate(new Date(expense.occurredAt), TIMESTAMP),
        formatCents(expense.amountCents),
        expense.kind,
        expense.description,
        joinPeople(expense.withWho),
      ]
        .map(escape)
        .join(','),
    )
  }
  return `${lines.join('\n')}\n`
}

function parseTimestamp(raw: string): number | null {
  const text = raw.trim().replace('T', ' ').replace(/Z$/, '')
  if (text === '') return null
  for (const candidate of [text, `${text}:00`, `${text} 00:00:00`]) {
    const parsed = parseDate(candidate, TIMESTAMP, new Date())
    if (isValid(parsed)) return parsed.getTime()
  }
  // A bare epoch-millis column is also accepted.
  const millis = Number.parseInt(text, 10)
  return Number.isSafeInteger(millis) && String(millis) === text ? millis : null
}

function parseRow(row: string[], newId: () => string, now: string): Expense | null {
  if (row.length < 4) return null
  if (row.every((cell) => cell.trim() === '')) return null

  const amountCents = parseAmount(row[2] ?? '')
  if (amountCents === null) return null

  const occurredAt = parseTimestamp(row[1] ?? '')
  if (occurredAt === null) return null

  const rawKind = (row[3] ?? '').trim().toUpperCase()
  if (rawKind !== 'NEED' && rawKind !== 'WANT') return null

  return {
    // A fresh id on every imported row: merging a backup into a database that
    // already has entries must add to it, never overwrite what is there.
    id: newId(),
    amountCents,
    kind: rawKind as Kind,
    description: (row[4] ?? '').trim(),
    occurredAt,
    withWho: dedupePeople(splitPeople(row[5] ?? '')),
    createdAt: now,
  }
}

export function importCsv(text: string, newId: () => string, now: string): ImportResult {
  const rows = parseRows(text)
  if (rows.length === 0) return { expenses: [], skipped: 0 }

  // Tolerate a file with or without a header line.
  const first = (rows[0] ?? []).map((cell) => cell.trim().toLowerCase())
  const body = first[0] === 'id' || first[2] === 'amount' ? rows.slice(1) : rows

  const expenses: Expense[] = []
  let skipped = 0
  for (const row of body) {
    const expense = parseRow(row, newId, now)
    if (expense === null) skipped++
    else expenses.push(expense)
  }
  return { expenses, skipped }
}

/** A small RFC 4180 reader: quoted fields, doubled quotes and embedded newlines. */
export function parseRows(text: string): string[][] {
  const rows: string[][] = []
  let row: string[] = []
  let field = ''
  let inQuotes = false
  let i = 0

  const endField = () => {
    row.push(field)
    field = ''
  }

  const endRow = () => {
    endField()
    if (row.length > 1 || (row[0] ?? '').trim() !== '') rows.push(row)
    row = []
  }

  while (i < text.length) {
    const c = text[i]
    if (inQuotes && c === '"' && text[i + 1] === '"') {
      field += '"'
      i++
    } else if (c === '"') {
      inQuotes = !inQuotes
    } else if (!inQuotes && c === ',') {
      endField()
    } else if (!inQuotes && (c === '\n' || c === '\r')) {
      if (c === '\r' && text[i + 1] === '\n') i++
      endRow()
    } else {
      field += c
    }
    i++
  }
  if (field !== '' || row.length > 0) endRow()
  return rows
}
